internal/driver/driver.go

```diff
@@ -0,0 +1,728 @@
+outloop:
+	for j := 0; j <= times; j++ {
+		items1 = items1[:0]
+		reqs1 = reqs1[:0] // clear array
+		if j >= times {
+			if remains > 0 {
+				items1 = items[times*batch_size : times*batch_size+remains]
+			} else {
+				break outloop // load complete, exit loop
+			}
+		} else {
+			items1 = items[j*batch_size : j*batch_size+batch_size]
+		}
+		// fmt.Println("reqs1: ", reqs1, "items1: ", items1)
+
+		// write data to S7
+		// fmt.Println("reqs len: ", len(items))
+		err = s7Client.Client.AGWriteMulti(items1, len(items1))
+		if err != nil {
+			s.lc.Errorf("AGWriteMulti Error: %s, reconnecting...", err)
+			s.mu.Lock()
+			s.s7Clients[deviceName] = s.NewS7Client(deviceName, protocols)
+			s.mu.Unlock()
+			return err
+		} else {
+			//fmt.Println(datas)
+		}
+	}
+	for _, err1 := range errors {
+		if err1 != "" {
+			s.lc.Errorf("S7 Client AGWriteMulti error: %s", err1)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Stop the protocol-specific DS code to shutdown gracefully, or
+// if the force parameter is 'true', immediately. The driver is responsible
+// for closing any in-use channels, including the channel used to send async
+// readings (if supported).
+func (s *Driver) Stop(force bool) error {
+
+	s.mu.Lock()
+	s.s7Clients = nil
+	s.mu.Unlock()
+
+	// Then Logging Client might not be initialized
+	if s.lc != nil {
+		s.lc.Debugf("Driver.Stop called: force=%v", force)
+	}
+	return nil
+}
+
+// AddDevice is a callback function that is invoked
+// when a new Device associated with this Device Service is added
+func (s *Driver) AddDevice(deviceName string, protocols map[string]models.ProtocolProperties, adminState models.AdminState) error {
+	s.lc.Debugf("a new Device is added: %s", deviceName)
+
+	s7Client := s.NewS7Client(deviceName, protocols)
+	if s7Client == nil {
+		s.lc.Errorf("failed to initialize S7 client for '%s' device, skipping this device.", deviceName)
+	}
+	s.mu.Lock()
+	s.s7Clients[deviceName] = s7Client
+	s.mu.Unlock()
+	return nil
+}
+
+// UpdateDevice is a callback function that is invoked
+// when a Device associated with this Device Service is updated
+func (s *Driver) UpdateDevice(deviceName string, protocols map[string]models.ProtocolProperties, adminState models.AdminState) error {
+	s.lc.Debugf("Device %s is updated", deviceName)
+	return nil
+}
+
+// RemoveDevice is a callback function that is invoked
+// when a Device associated with this Device Service is removed
+func (s *Driver) RemoveDevice(deviceName string, protocols map[string]models.ProtocolProperties) error {
+	s.lc.Debugf("Device %s is removed", deviceName)
+	s.mu.Lock()
+	s.s7Clients[deviceName] = nil
+	s.mu.Unlock()
+	return nil
+}
+
+func (s *Driver) ValidateDevice(device models.Device) error {
+
+	return nil
+}
+
+// Create S7Client by 'Device' definition
+func (s *Driver) NewS7Client(deviceName string, protocol map[string]models.ProtocolProperties) *S7Client {
+
+	pp := protocol[Protocol]
+	var errt error
+	host, errt := cast.ToStringE(pp["Host"])
```

Comment:
These error checks should be in the ValidateDevice function and then are not needed here since a device added was already validated.